pkg/cloudprovider/provider/packet/provider.go

```diff
@@ -0,0 +1,420 @@
+	Facilities   []string
+	OS           string
+	Tags         []string
+}
+
+type provider struct {
+	configVarResolver *providerconfig.ConfigVarResolver
+}
+
+func (p *provider) getConfig(s v1alpha1.ProviderSpec) (*Config, *providerconfig.Config, error) {
+	if s.Value == nil {
+		return nil, nil, fmt.Errorf("machine.spec.providerconfig.value is nil")
+	}
+	pconfig := providerconfig.Config{}
+	err := json.Unmarshal(s.Value.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rawConfig := RawConfig{}
+	if err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig); err != nil {
+		return nil, nil, err
+	}
+
+	c := Config{}
+	c.APIKey, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.APIKey, "PACKET_API_KEY")
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"apiKey\" field, error = %v", err)
+	}
+	c.InstanceType, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.InstanceType)
+	if err != nil {
+		return nil, nil, err
+	}
+	for i, tag := range rawConfig.Tags {
+		tagValue, err := p.configVarResolver.GetConfigVarStringValue(tag)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get the %d value of \"tags\" field, error = %v", i, err)
+		}
+		c.Tags = append(c.Tags, tagValue)
+	}
+	for i, facility := range rawConfig.Facilities {
+		facilityValue, err := p.configVarResolver.GetConfigVarStringValue(facility)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get the %d value of \"facilities\" field, error = %v", i, err)
+		}
+		c.Facilities = append(c.Facilities, facilityValue)
+	}
+	return &c, &pconfig, err
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
+	c, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.APIKey == "" {
+		return errors.New("apiKey is missing")
+	}
+
+	_, err = getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	client := getClient(c.APIKey)
+
+	if len(c.Facilities) == 0 || c.Facilities[0] == "" {
+		return fmt.Errorf("must have at least one non-blank facility")
+	}
+
+	// get all valid facilities
+	facilities, _, err := client.Facilities.List(nil)
+	if err != nil {
+		return fmt.Errorf("failed to list facilities: %v", err)
+	}
+	// ensure our requested facilities are in those facilities
+	if missingFacilities := itemsNotInList(facilityProp(facilities, "Code"), c.Facilities); len(missingFacilities) > 0 {
+		return fmt.Errorf("unknown facilities: %s", strings.Join(missingFacilities, ","))
+	}
+
+	// get all valid plans a.k.a. instance types
+	plans, _, err := client.Plans.List(nil)
+	if err != nil {
+		return fmt.Errorf("failed to list instance types / plans: %v", err)
+	}
+	// ensure our requested plan is in those plans
+	if missingPlans := itemsNotInList(planProp(plans, "Slug"), []string{c.InstanceType}); len(missingPlans) > 0 {
```

Comment:
Same here, what about just getting the one we are interested in?

Reply:
Same for `Plans` as for `Facilities`. It is a simple `N(1)` op, and there is no `GET` for it.